discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,220 @@
+package zookeeper
+
+import (
+	"encoding/json"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/utils"
+	"github.com/samuel/go-zookeeper/zk"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/joyent/containerpilot/discovery"
+)
+
+func init() {
+	discovery.RegisterBackend("zookeeper", ConfigHook)
+}
+
+// ZooKeeper is a wrapper ZooKeeper connection. It also stores the
+// prefix under which ContainerPilot nodes will be registered.
+type ZooKeeper struct {
+	Connection *zk.Conn
+	Prefix     string
+}
+
+// ServiceNode is the serializable form of a ZooKeeper service record
+type ServiceNode struct {
+	ID      string   `json:"id"`
+	Name    string   `json:"name"`
+	Address string   `json:"address"`
+	Port    int      `json:"port"`
+	Tags    []string `json:"tags"`
+}
+
+type zookeeperRawConfig struct {
+	Endpoints interface{} `mapstructure:"endpoints"`
+	Prefix    string      `mapstructure:"prefix"`
+}
+
+// ConfigHook is the hook to register with the ZooKeeper backend
+func ConfigHook(raw interface{}) (discovery.ServiceBackend, error) {
+	return NewZooKeeperConfig(raw)
+}
+
+// NewZooKeeperConfig creates a new service discovery backend for zookeeper
+func NewZooKeeperConfig(raw interface{}) (*ZooKeeper, error) {
+	zookeeper := &ZooKeeper{Prefix: "/containerpilot"}
+	config := &struct {
+		Address string `mapstructure:"address"`
+	}{}
+	if err := utils.DecodeRaw(raw, &config); err != nil {
+		return nil, err
+	}
+	c, _, err := zk.Connect([]string{config.Address}, time.Second)
+	if err != nil {
+		return nil, err
+	}
+	zookeeper.Connection = c
+	return zookeeper, nil
+}
+
+// Deregister removes this instance from the registry
+func (c *ZooKeeper) Deregister(service *discovery.ServiceDefinition) {
+	c.Connection.Delete(c.getNodeKey(service), -1)
+}
+
+// MarkForMaintenance removes this instance from the registry
+func (c *ZooKeeper) MarkForMaintenance(service *discovery.ServiceDefinition) {
+	c.Deregister(service)
+}
+
+// SendHeartbeat refreshes the TTL of this associated zookeeper node
```

Comment:
We don't have an actual TTL, so we should adjust this comment so that it isn't misleading.